edb/edgeql-parser/src/parser.rs

```diff
@@ -0,0 +1,572 @@
+use indexmap::IndexMap;
+
+use crate::helpers::quote_name;
+use crate::keywords::Keyword;
+use crate::position::Span;
+use crate::tokenizer::{Token, Value, Error, Kind};
+
+pub struct Context<'s> {
+    spec: &'s Spec,
+    arena: bumpalo::Bump,
+}
+
+impl<'s> Context<'s> {
+    pub fn new(spec: &'s Spec) -> Self {
+        let arena = bumpalo::Bump::new();
+        Context { spec, arena }
+    }
+}
+
+pub fn parse<'a>(input: &'a [Terminal], ctx: &'a Context) -> (Option<&'a CSTNode<'a>>, Vec<Error>) {
+    let stack_top = ctx.arena.alloc(StackNode {
+        parent: None,
+        state: 0,
+        value: CSTNode::Empty,
+    });
+    let initial_track = Parser {
+        stack_top,
+        error_cost: 0,
+        node_count: 0,
+        can_recover: true,
+        errors: Vec::new(),
+    };
+
+    // append EIO
+    let end = input.last().map(|t| t.span.end).unwrap_or_default();
+    let eio: &Terminal = ctx.arena.alloc(Terminal {
```

Comment:
Putting `Terminal`s in a bumpalo arena will leak `text` and `value`, I think, since bumpalo doesn't run `Drop`s. (This is done here and in `new_token_for_injection`, at least).

One answer might be to have a `Vec` in the `Context` for storing the `Terminal`s, and then we can just take pointers off that

Reply:
We only have an unmutable reference to Context, so we couldn't add new terminals to the vec.

But I can wrap this into a buffalo::Box that drops when an arena drops.

Reply:
Hmmm, I don't think that's what `buffalo::Box` does, looking at it? I think it creates a box that *when dropped*, runs the interior Drop. But if you put the `buffalo::Box` *in the arena*, that never happens?

Reply:
Maybe https://docs.rs/append-only-vec/latest/append_only_vec/ is what we need?

Reply:
You're right, I misinterpreted this:

> However: [bumpalo::boxed::Box<T>](https://docs.rs/bumpalo/latest/bumpalo/boxed/struct.Box.html) can be used to wrap T values allocated in the Bump arena, and calls T's Drop implementation when the Box<T> wrapper goes out of scope. This is similar to how [std::boxed::Box](https://doc.rust-lang.org/std/boxed/struct.Box.html) works, except without deallocating its backing memory.

Here are bumpalo docs talking about this:

https://docs.rs/bumpalo/latest/bumpalo/struct.Bump.html#no-drops